### PR TITLE
JSONRPCRequest-aware RPCHelpMan

### PR DESCRIPTION
**Pull request by kallewoof**
Every single RPC call has a helper-section at the start, which throws a help string if the user asks for help or if the user provided too few/many arguments.

```C++
const RPCHelpMan help{...};
if (request.fHelp || !help.IsValidNumArgs(request.params.size())) {
    throw std::runtime_error(help.ToString());
}
```

or (older version)

```C++
if (request.fHelp || request.params.size() < min || request.params.size() > max)
    throw std::runtime_error(
        RPCHelpMan{...}.ToString()
    );
```

It seems like an obvious improvement, and less copy-pasting, to make `RPCHelpMan` aware of `JSONRPCRequest`, and to let it handle the checks instead. Both of the above become

```C++
RPCHelpMan{...}.Check(request);
```

which means we save roughly 3 lines per RPC command, and the `RPCHelpMan` instance is never referenced afterwards, so the approach is a tiny fraction cleaner.

This is a complete update, sans a few special case locations that had special rules. 623 lines turn into 284 (which includes the addition to `RPCHelpMan`).

### CODE REVIEW

**Comment by sipa** [comment]
Concept ACK, though it seems more appropriate that the RPC help object would gain a method the check RPC arguments, than the RPC handling code to be aware of RPC help.

Also, if these lines are all touched anyway, could you add a scripted diff to fix their indentation?

**Comment by promag** [comment]
Agree with @sipa, although the request already knows about `fHelp`.

**Comment by kallewoof** [comment]
@sipa I initially added one, but it needs to know about `JSONRPCRequest` to fetch the `fHelp` and `params.size()`, but if I `#include <rpc/server.h>`, the linters detect a circular dependency `util -> server -> util`. I definitely think this would be better in `RPCHelpMan`, but alas..

I would love to fix the indentation, but not sure how a scripted-diff would be used in this case. (I started working on that, with a sample commit at https://github.com/kallewoof/bitcoin/commit/b9694578fa6401508e18761954ded33ed053df94 in https://github.com/kallewoof/bitcoin/commits/json-aware-helper-whitespaced-sample but got stuck on how to do scripted-diff part).

**Comment by fanquake** [comment]
Concept ACK - nice cleanup / deduplication.

**Comment by kallewoof** [comment]
@sipa @promag I realized I could inline the helper method and implement it after `JSONRPCRequest` to get around linker issues, so this is now a `RPCHelpMan` method, as you recommended.

**Comment by promag** [comment]
@kallewoof see dfdf7f213cd84756f8f17bb3c7cb25a1ee7989e2 for a scripted diff, produced by running on your 31599d8:
```sh
git diff -U0 HEAD~3.. | ./contrib/devtools/clang-format-diff.py -p0 -i -v
```

**Comment by kallewoof** [comment]
@promag I had no idea you could do that. Thanks, very cool!

File: src/rpc/util.h [hunk 1]
```diff
@@ -236,6 +238,7 @@ class RPCHelpMan
     std::string ToString() const;
     /** If the supplied number of args is neither too small nor too high */
     bool IsValidNumArgs(size_t num_args) const;
+    inline void Check(const JSONRPCRequest& request) const;
```
Comment (MarcoFalke) [inline src/rpc/util.h]: in commit 2512d3d2e2f2a2b891f9a3baa1a1309edaf41e01:

Missing doc string and could make `IsValidNumArgs` private?

**Review by maflcko — COMMENTED** [review]
Tend to NACK on fixing the code style in lines that are not touched otherwise. This breaks `git blame` and all cherry-pick/rebase/merge operations by default (see the list of conflicts).

**Comment by promag** [comment]
Makes sense @MarcoFalke, not worth it.

File: src/rpc/util.h (hunk 1 above)
Comment (kallewoof) [reply src/rpc/util.h]: Done!

**Comment by kallewoof** [comment]
@MarcoFalke 

> Tend to NACK on fixing the code style in lines that are not touched otherwise. This breaks `git blame` and all cherry-pick/rebase/merge operations by default (see the list of conflicts).

I would normally agree, but the RPC code is out-of-this-world awful with indentation. I think it may be worth to run through it once, especially since the review effort required is near-nil.

**Comment by MarcoFalke** [comment]
> especially since the review effort required is near-nil.

Just because something is easy to review doesn't mean it should be done. Beside the things it breaks that I mentioned previously, it will also conflict with a bunch of my local branches, where I restructure the help to make it more consistent and compile-time-safe (use compile time checks and compile-time generated help). Those changes are going to actually touch the lines, so a reformat could make sense at that point in time (maybe).

Also, I'd say it is not clear if the rpc help *must* live in the scope of the rpc method. Maybe the help should live at the same scope as the rpc method? I'd say unclear right now, but if it is ever going to change, it would need another reformat. Doing another project-wide break of `git blame/cherry-pick/merge/rebase`

I think you underestimate the cost of merge conflicts. If you feel strong about the reformat, it should probably be split up into a separate pull request to be evaluated on its own. Otherwise, my NACK will still hold on this pull.

**Comment by kallewoof** [comment]
Understood. Would have been a nice clean up but I'm not gonna lose sleep over it. :) Dropped the whitespace commit.

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -302,7 +302,7 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
                 "    https://github.com/bitcoin/bips/blob/master/bip-0009.mediawiki#getblocktemplate_changes\n"
                 "    https://github.com/bitcoin/bips/blob/master/bip-0145.mediawiki\n",
                 {
-                    {"template_request", RPCArg::Type::OBJ, RPCArg::Optional::NO, "A json object in the following spec",
+                    {"template_request", RPCArg::Type::OBJ, "{\"mode\":\"template\"}", "A json object in the following spec",
```
Comment (MarcoFalke) [inline src/rpc/mining.cpp]: Does this need backport? Should be a separate pull request, since behavior changes should not be mixed with refactoring.

Comment (kallewoof) [reply src/rpc/mining.cpp]: It may be wrong doc-wise, but tests break if you make this mandatory. And it *is* optional in the code.

Comment (MarcoFalke) [reply src/rpc/mining.cpp]: Which tests break?

Comment (kallewoof) [reply src/rpc/mining.cpp]: minig_basic.py (see above log)

Comment (MarcoFalke) [reply src/rpc/mining.cpp]: Ah, so the default is not `{\"mode\":\"template\"}`, but the empty dict `{}`? The help text for mode says "This must be set to \"template\", \"proposal\" (see BIP 23), **or omitted**" (emphasis mine)

Comment (kallewoof) [reply src/rpc/mining.cpp]: If it can be omitted, `Optional::NO` is probably wrong, then. I can change the default to simply say `{}`, but the code essentially acts as if it was `{"mode":"template"}`..

**Comment by kallewoof** [comment]
@MarcoFalke: Updated `getblocktemplate` to display default as `{}` rather than `{"mode":"template"}`.

File: src/rpc/server.h [hunk 3]
```diff
@@ -42,6 +44,10 @@ class JSONRPCRequest
     void parse(const UniValue& valRequest);
 };
 
+inline void RPCHelpMan::Check(const JSONRPCRequest& request) const {
+    if (request.fHelp || !IsValidNumArgs(request.params.size())) throw std::runtime_error(ToString());
+}
```
Comment (MarcoFalke) [inline src/rpc/server.h]: In commit 09e176a958  Make the RPCHelpMan aware of JSONRPCRequest and add Check() helper


Looks like you added the implementation to an unrelated header file (should be in rpc/util.cpp).

You can solve the circular dependency by moving JSONRPCRequest to rpc/protocol.h

Comment (kallewoof) [reply src/rpc/server.h]: That sounds like a good solution, yeah. Thanks, doing that.

File: src/rpc/server.cpp [hunk 4]
```diff
@@ -7,7 +7,6 @@
 
 #include <fs.h>
 #include <key_io.h>
-#include <rpc/util.h>
```
Comment (MarcoFalke) [inline src/rpc/server.cpp]: ? :eyes:

Comment (kallewoof) [reply src/rpc/server.cpp]: Sigh. I moved it to .h because it needed RPCHelpMan at one point. Fixed!

**Review by promag — COMMENTED** [review]
ACK 36ad069e8a2641fc7a04af3b924ca05b8b91b6f0.

File: src/rpc/protocol.h [hunk 5]
```diff
@@ -106,4 +106,19 @@ void DeleteAuthCookie();
 /** Parse JSON-RPC batch reply into a vector */
 std::vector<UniValue> JSONRPCProcessBatchReply(const UniValue &in, size_t num);
 
+class JSONRPCRequest
```
Comment (laanwj) [inline src/rpc/protocol.h]: I'm not 100% convinced that `protocol.h` is the right place for this architecturally, as it's supposed to be functions and data structures shared between RPC client and server, and `JSONRPCRequest` is internal to the server. I understand the rationale, though.

**Review by laanwj — COMMENTED** [review]
Concept ACK on the approach, I think making this checking data-driven is, overall, a good idea (it enforces consistency), and this seems to be a good way to go about it.
I do think this is slightly risky, as it gets rid of a lot of special-case code, is our test coverage of RPC good enough to be confident about changing this for all methods at once?

File: src/rpc/protocol.h (hunk 5 above)
Comment (kallewoof) [reply src/rpc/protocol.h]: I think you're right, but I also think we have stuff in `protocol` that does not belong there to begin with. I made an attempt at cleaning this up by splitting the 'request' stuff into a separate file. This moves stuff from `server.cpp` and `server.h` and `protocol.h` into `request.h`, and `protocol.cpp` is replaced by `request.cpp`.

**Comment by kallewoof** [comment]
@laanwj

> I do think this is slightly risky, as it gets rid of a lot of special-case code, is our test coverage of RPC good enough to be confident about changing this for all methods at once?

I'm not sure about "a lot" of special-case code, to be honest. There were a few exceptions but I mostly left special case stuff alone.

**Comment by laanwj** [comment]
> I'm not sure about "a lot" of special-case code, to be honest. There were a few exceptions but I mostly left special case stuff alone.

OK, I just wanted to bring it up.

File: src/rpc/util.h [hunk 6]
```diff
@@ -243,6 +243,13 @@ class RPCHelpMan
     std::string ToString() const;
     /** If the supplied number of args is neither too small nor too high */
     bool IsValidNumArgs(size_t num_args) const;
+    /**
+     * Check if the given request is valid according to this command or if
+     * the user is asking for help information, and throw help when appropriate.
+     */
+    inline void Check(const JSONRPCRequest& request) const {
+        if (request.fHelp || !IsValidNumArgs(request.params.size())) throw std::runtime_error(ToString());
```
Comment (promag) [inline src/rpc/util.h]: 09f0bf5f1a5a50168d64c488c2b8972774711a98

nit, add `{ }`.

**Review by promag — COMMENTED** [review]
ACK d9c41cc4cb, github commit sort ftw.. Yet another simple tidy up to RPC documentation. Renaming `rpc/protocol.{h.cpp}` LGTM - thanks to @laanwj. Also confirmed moved code.

File: src/rpc/util.h (hunk 6 above)
Comment (kallewoof) [reply src/rpc/util.h]: Added.

**Comment by kallewoof** [comment]
@promag Thanks for the review! Note that I only renamed `protocol.cpp`, and moved some parts out of `protocol.h` into `request.h` -- `protocol.h` is still there, just smaller.

**Comment by laanwj** [comment]
code rview and lightly tested ACK b6fb617aaaad5f9cdd7f2ad2825b253ca792055d

**Comment by MarcoFalke** [comment]
>     I do think this is slightly risky, as it gets rid of a lot of special-case code, is our test coverage of RPC good enough to be confident about changing this for all methods at once?


I checked all call-sites as part of
*  rpc: Actually throw help when passed invalid number of params #15401 

so this *should* only be refactoring. Though, you are right, most of it is not covered by tests.

**Comment by MarcoFalke** [comment]
ACK b6fb617aaa, looked at the diff, verified move-only where applicable

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK b6fb617aaa, looked at the diff, verified move-only where applicable
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhX0AwAtwImLc9meZaRYasAtXwhpkTQpZ9bC0l6hK6uvUGSr4N8n3v63bezavGZ
pbddzccHclbs5gAtbJMsWkFmR7vKbfDgZi02MZLQqLwc5r+X8/rJdgscfFhMeHNy
WCQ8ZCJy5VAC5GA6SfshB8VWK+oQpx9GHCKc5OEMFRr7Sc8go0osRxWbMaNuJQdg
RGgzZ0/e2FS17czydBb5O3k2XTawUUfIa49zFq69kxSntpJdnmMJrqbAg8l0cKDK
OYd/HkWIVgx1LBtTBmLEF0sKuymhmqMfPbEoFOUXYqiGSTaQsZeS9cJDpc+Cy8Qa
+icW5+C23plWHC/+N89AeE6g3zcdxG1P1pIRP0YfRjQMx1lKxCuYeCep8vr2pNw7
X+SpZaoyRJWaoQnL4mPjtGP/7n+QXXjYj2MCq8yNz2uPzyqrvcfq7ZaP+IW5SBmY
Yl+J9K6FwTkneVkhDMjYMzoJLDwDH4yVrgChwMHYN7HEVY2M0WZhsSE9/HDgscqg
jEii4j6x
=rpAy
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `890e34cddd2f763e5555c79809bdd129b142d59574aa8799ab8f1a06a0c076c6  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108890e34cddd2f763e5555c79809bdd129b142d59574aa8799ab8f1a06a0c076c6f010072a813a3df65cdc881c75f09a50b42e08fff010114743a5bbd310ab67ed3d4c117a259d08f1045d2523c5f0087c7f52030826e53c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01073ad38255050f684f6cb2f8a5ccd381908f1045d2523c5f008ec69f97b39b9c65f0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c8a99f4e8cc55e50b35ee56f369ca0d708f1045d2523c5f0089b217a6d72a6df570083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e912159d6c0af05c39814639dc034aef08f02038ee9ff4801acd436848509bec4a0c2716850e84afde9f089f01529249e3414108f02082fc44151b5c5355b786ba61229a076c131a9d72224a048148ddd8b60b15bccc08f1045d2523c5f0087dbcec818903be650083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

**Review by CathCathub — APPROVED** [review]
